### ISSUE TITLE
Unable to load `ring.defaults.middleware`

### ISSUE THREAD

**Issue by metasoarous** [cljdoc/cljdoc#404]
Getting this issue trying to build docs for Oz:

https://app.circleci.com/pipelines/github/cljdoc/builder/7952/workflows/f86543e9-4c52-41a2-9b2a-8d6561f944d2/jobs/24329/steps

Honing in:

```
Caused by: java.io.FileNotFoundException: Could not locate ring/middleware/defaults__init.class, ring/middleware/defaults.clj or ring/middleware/defaults.cljc on classpath.
```

I have `ring/ring-defaults {:mvn/version "0.3.2"}` (the latest) in my deps, so assuming this is a conflict with something in cljdoc? Is it possible cljdoc is using an old version of this?

Thanks!

**Comment by SevereOverfl0w** [comment]
```
org.clojure/clojure 1.10.0
  org.clojure/core.specs.alpha 0.2.44
  org.clojure/spec.alpha 0.2.176
org.clojure/clojurescript 1.10.773
  org.clojure/data.json 0.2.6
  org.clojure/google-closure-library 0.0-20191016-6ae1f72f
    org.clojure/google-closure-library-third-party 0.0-20191016-6ae1f72f
  com.cognitect/transit-clj 0.8.309
    com.cognitect/transit-java 0.8.332
      commons-codec/commons-codec 1.10
      com.fasterxml.jackson.core/jackson-core 2.8.7
      org.msgpack/msgpack 0.6.12
        com.googlecode.json-simple/json-simple 1.1.1
        org.javassist/javassist 3.18.1-GA
  org.clojure/tools.reader 1.3.2
  com.google.javascript/closure-compiler-unshaded v20200315
    com.google.elemental2/elemental2-core 1.0.0-RC1
    com.google.errorprone/error_prone_annotations 2.3.1
    com.google.jsinterop/jsinterop-annotations 1.0.2
    com.google.javascript/closure-compiler-externs v20200315
    com.google.guava/guava 25.1-jre
      org.codehaus.mojo/animal-sniffer-annotations 1.14
      com.google.j2objc/j2objc-annotations 1.1
      org.checkerframework/checker-qual 2.0.0
      com.google.code.findbugs/jsr305 3.0.2
    args4j/args4j 2.0.26
    com.google.jsinterop/base 1.0.0
    com.google.re2j/re2j 1.3
    com.google.protobuf/protobuf-java 3.11.1
    com.google.code.gson/gson 2.7
cljdoc-analyzer/metagetta /home/overfl0w/src/github.com/cljdoc/cljdoc-analyzer/modules/metagetta
  org.clojure/tools.namespace 0.3.1
    org.clojure/java.classpath 0.3.0
javax.servlet/javax.servlet-api 4.0.1
metasoarous/oz /tmp/cljdoc-metasoarous-oz-1.6.0-alpha714516161777634023385/downloaded.jar
```

Looks like no deps are being resolved for your lib.  Looking deeper, this seems to be because your jar is lacking a META-INF/pom.xml file that tools deps will use to resolve dependencies when using a local jar.

cljdoc should handle this really, there's some tricks that could be done to make it happen.

**Comment by SevereOverfl0w** [comment]
Ah, you're using pack!  Yeah, my guide for tdeps & jars was unfortunately minimal.  Martin has pestered me about this before.

Simple fix is to do something like:

```
mkdir -p foo/META-INF
cp pom.xml foo/META-INF/
clojure -m mach.pack…
```

I still think cljdoc should learn how to handle this by itself though.

**Comment by SevereOverfl0w** [comment]
You will need to add the pom.xml for other reasons, e.g. for a link to your github.

**Comment by metasoarous** [comment]
Hi @SevereOverfl0w! Thanks for looking into this, and for the advise!

I gave this a whirl, but am still having the same problem:

https://app.circleci.com/pipelines/github/cljdoc/builder/7964/workflows/1a14f8f6-5bd1-4724-bc35-616029fe2e80/jobs/24341/steps

Here's my `Makefile`: https://github.com/metasoarous/oz/blob/master/Makefile

I wasn't sure what `foo` should be, if not the project root. But also, something seems off to me about the order here. I only run `clojure -Spom` _after_ running `clojure -A:pack ...`, so this would seem to miss updates on the `pom.xml`. Feels like I'm misunderstanding something here. Please let me know if you see where I'm going wrong.

Thanks again!

**Comment by SevereOverfl0w** [comment]
@metasoarous I forgot to do `-e foo` in my original example, sorry!

You need a parent directory to use with -e. foo could be pretty much anything. I can take a deeper look when I'm back at a keyboard

**Comment by metasoarous** [comment]
Hello again. Coming back around to this because I'm hoping to release the first non-alpha of Oz in a while pretty soon, and getting docs going is one of just a few things holding me up.

I believe I tried what you suggested @SevereOverfl0w (adding `-e`) and that didn't seem to help (though this was a while ago now, so don't remember the details). I'm feeling a bit confused about a few things:
* At the very least, my github README is showing up on the cljdoc page, so it seems to be able to get _some_ information from my `pom.xml`; Why aren't the dependencies coming along for the ride?
* Why does the `META_INF` have be under a subdirectory? This seems strange, and makes it unclear how cljdoc would even find that content then, since my understanding is that the expectation would be for `META_INF` to be in the root of the project.
* Should `clojure -Spom` come _before_ or _after_ `clojure -A:pack -e ...`?

You also said

> cljdoc should handle this really, there's some tricks that could be done to make it happen.

Is a quick fix for this in the cards as part of cljdoc? If not would it be easier for me to switch to [depstar](https://github.com/seancorfield/depstar)? Are there any complete examples of this you could point me to?

Thanks again for all your assistance!

**Comment by SevereOverfl0w** [comment]
There's an open PR that's almost ready to go.  Just waiting on one last check :)

Then you'll have nothing to do!

18 Nov 2020 00:10:09 Christopher Small <notifications@github.com>:

> Hello again. Coming back around to this because I'm hoping to release the first non-alpha of Oz in a while pretty soon, and getting docs going is one of just a few things holding me up.
> 
> I believe I tried what you suggested @SevereOverfl0w[https://github.com/SevereOverfl0w] (adding -e) and that didn't seem to help (though this was a while ago now, so don't remember the details). I'm feeling a bit confused about a few things:
> 
* > At the very least, my github README is showing up on the cljdoc page, so it seems to be able to get some information from my pom.xml; Why aren't the dependencies coming along for the ride?
* > Why does the META_INF have be under a subdirectory? This seems strange, and makes it unclear how cljdoc would even find that content then, since my understanding is that the expectation would be for META_INF to be in the root of the project.
* > Should clojure -Spom come before or after clojure -A:pack -e ...?

**Comment by SevereOverfl0w** [comment]
@metasoarous This is now fixed, looks like you've got a new bug though (don't think I caused this one!) https://app.circleci.com/pipelines/github/cljdoc/builder/12203/workflows/960f0cf7-f480-45ca-a65e-f3533d9b0d3c/jobs/28580

You need to add clojupyter as a provided dependency in your pom.xml, see https://github.com/cljdoc/cljdoc/blob/03a4a5defbef750c71ac971aa5e4435a28deffea/doc/userguide/for-library-authors.adoc#getting-dependencies-right

**Comment by metasoarous** [comment]
Hi @SevereOverfl0w; Thanks so much for fixing that! As well as for pointing out the next hurdle.

Unfortunately, it doesn't seem to be possible to add dependencies with `:scope "provided"` to `deps.edn`, as it essentially just ignores them. Worse, if I manually put such a dep in the `pom.xml`, it simply removes it.  I imagine I could write a script that manually munges this in after the pom has been built, but that feels pretty hacky.

Is there no way we could support some bit of specification in the `doc/cljdoc.edn` config that would let us specify aliases which need to be looked at for additional dependencies?

**Comment by SevereOverfl0w** [comment]
@metasoarous You're right about both deps.edn and pom.xml for this case.  After the initial generation, I've just been maintaining my own `pom.xml` manually, which sucks a little, agreed.

Not sure if the code is amenable to that change off the top of my head.  Might also be worth a feature request to deps.edn to ignore provided deps and not remove them perhaps?  Dunno.

Might be a simple program to write separately too, I don't believe -Spom is too big.

**Comment by metasoarous** [comment]
Thanks for the follow up @SevereOverfl0w.

Yeah, I meant to mention in my last message, but I did actually suggest that here: https://ask.clojure.org/index.php/9110/scope-in-deps-edn-should-be-added-and-not-deleted-from-pom-xml?show=9833

I'm attempting to write a little babashka script for this, but it's not fun (xml :vomiting_face:).

Thanks again

**Comment by metasoarous** [comment]
Update: I went ahead and wrote a babashka script for glomming on the provided scope dependency. So through that hurdle.

But now, I'm getting this:

https://app.circleci.com/pipelines/github/cljdoc/builder/12263/workflows/744f81e5-eba7-43db-b1e6-2854901f0179/jobs/28640

In short, it looks like it's having trouble processing the cljs dependencies. Searching around a bit I'm finding relevant issues regarding this, but am a bit done for the night. Please let me know if you are immediately able to spot what's amiss; Otherwise, I'll take another look (tomorrow evening hopefully).

Thanks again!

**Comment by metasoarous** [comment]
Since the original issue I brought up here has been resolved, I'm going to close this and start a new issue for the cljs issue mentioned above.

Thanks again for all your help!
